bfvmm/src/hve/arch/intel_x64/ept/violation.cpp

```diff
@@ -0,0 +1,212 @@
+        exit_reason::basic_exit_reason::ept_violation,
+        ::handler_delegate_t::create<violation, &violation::handle>(this)
+    );
+
+    this->add_read_handler(handler_delegate_t::create<default_handler>());
+    this->add_write_handler(handler_delegate_t::create<default_handler>());
+    this->add_execute_handler(handler_delegate_t::create<default_handler>());
+}
+
+violation::~violation()
+{
+    if (!ndebug && m_log_enabled) {
+        dump_log();
+    }
+}
+
+void
+violation::add_read_handler(handler_delegate_t &&d)
+{ m_read_handlers.push_front(std::move(d)); }
+
+void
+violation::add_write_handler(handler_delegate_t &&d)
+{ m_write_handlers.push_front(std::move(d)); }
+
+void
+violation::add_execute_handler(handler_delegate_t &&d)
+{ m_execute_handlers.push_front(std::move(d)); }
+
+void
+violation::dump_log()
+{
+    if (!m_read_log.empty()) {
+        bfdebug_transaction(0, [&](std::string * msg) {
+            bfdebug_lnbr(0, msg);
+            bfdebug_info(0, "ept read violation log", msg);
+            bfdebug_brk2(0, msg);
+
+            for (const auto &record : m_read_log) {
+                bfdebug_subnhex(0, "guest virtual address", record.gva, msg);
+                bfdebug_subnhex(0, "guest physical address", record.gpa, msg);
+            }
+
+            bfdebug_lnbr(0, msg);
+        });
+    }
+
+    if (!m_write_log.empty()) {
+        bfdebug_transaction(0, [&](std::string * msg) {
+            bfdebug_lnbr(0, msg);
+            bfdebug_info(0, "ept write violation log", msg);
+            bfdebug_brk2(0, msg);
+
+            for (const auto &record : m_write_log) {
+                bfdebug_subnhex(0, "guest virtual address", record.gva, msg);
+                bfdebug_subnhex(0, "guest physical address", record.gpa, msg);
+            }
+
+            bfdebug_lnbr(0, msg);
+        });
+    }
+
+    if (!m_execute_log.empty()) {
+        bfdebug_transaction(0, [&](std::string * msg) {
+            bfdebug_lnbr(0, msg);
+            bfdebug_info(0, "ept execute violation log", msg);
+            bfdebug_brk2(0, msg);
+
+            for (const auto &record : m_execute_log) {
+                bfdebug_subnhex(0, "guest virtual address", record.gva, msg);
+                bfdebug_subnhex(0, "guest physical address", record.gpa, msg);
+            }
+
+            bfdebug_lnbr(0, msg);
+        });
+    }
+}
+
+bool
+violation::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    namespace ept_violation = vmcs_n::exit_qualification::ept_violation;
+
+    struct info_t info = {
+        vmcs_n::guest_linear_address::get(),
+        vmcs_n::guest_physical_address::get(),
+        false
+    };
+
+    if (ept_violation::data_read::is_enabled()) {
+        return handle_read(vmcs, info);
+    }
+
+    if (ept_violation::data_write::is_enabled()) {
+        return handle_write(vmcs, info);
+    }
+
+    if (ept_violation::instruction_fetch::is_enabled()) {
+        return handle_execute(vmcs, info);
+    }
+
```

Comment:
In addition to the exception, I think a dump of the info struct's fields would be helpful too